lib/Sema/CSGen.cpp

```diff
@@ -1712,6 +1712,38 @@ namespace {
       return CS.getType(expr->getSubExpr());
     }
 
+    Type visitDotSelfExpr(DotSelfExpr *expr) {
+      auto result = visitIdentityExpr(expr);
+
+      // Dig through the result to see if warnings are needed.
+      auto metaTy = result->getAs<AnyMetatypeType>();
+      if (metaTy == nullptr)
+        return result;
+      if (!metaTy->getInstanceType()->is<AnyMetatypeType>())
+        return result;
+      auto typeExpr = dyn_cast<TypeExpr>(expr->getSubExpr());
+      if (typeExpr == nullptr)
+        return result;
+      auto typeRepr = typeExpr->getTypeRepr();
+      if (isa<TupleTypeRepr>(typeRepr))
+        return result;
+
+      auto &diags = CS.getASTContext().Diags;
+
+      if (auto metaRepr = dyn_cast<MetatypeTypeRepr>(typeRepr)) {
+        diags.diagnose(expr->getSelfLoc(), diag::metatype_self)
+          .fixItRemove(metaRepr->getMetaLoc());
+      } else {
+        diags.diagnose(expr->getSelfLoc(), diag::metatype_self);
+      }
+
+      diags.diagnose(typeExpr->getStartLoc(), diag::metatype_self_silence)
+        .fixItInsert(typeExpr->getStartLoc(), "(")
+        .fixItInsertAfter(typeExpr->getEndLoc(), ")");
+
+      return result;
```

Comment:
What if the user provides an explicit type annotation? For example:

```swift
let foo: AnyObject.Type.Protocol = AnyObject.Type.self
```

Should the warning still be emitted?

Reply:
I think so. Meta-metatypes (and meta-meta-metatypes etc, etc) are almost completely devoid of utility in the user-facing language and likely to remain that way. If the user tries to use them, they're probably wandering into the weeds and a warning is appropriate, even if explicit types are used.

Reply:
If I might get opinionated, I'd deprecate meta-metatypes (and meta-meta-metatypes, etc, etc) in the user-facing language. They don't have any members, nor can they be defined or extended. They also don't really serve a purpose other than to expose how the compiler models metatypes internally.